### ISSUE TITLE
Untokenized Regex Pattern Matching using Phrasematcher 

### ISSUE THREAD

**Issue by sohilsshah91** [explosion/spaCy#3796]
Hi ,
I wanted to try regex based multiple patterns getting different custom named entities but in an untokenized fashion. I have read about PhraseMatcher but I read we do not have custom regex patterns in PhraseMatcher.

Is there any way I can connect the 2? This is the regex I am checking and want to tag to a new entity:
`re.compile(r'(\d+\s*mg/m\d*` to \d+\s*mg/m\d*)|(\d+ to `\d+\s*mg/m\d*)|(\d+\s*mg/m\d*)|(\d+\s*\w+\s*units/m\d*)')`

I have tried using references from [Github Reference](https://github.com/dado0583/JupyterNotebooks/blob/master/SpacyExamples/AddCustomEntityMatcher.ipynb),  [Regular expressions in older versions part](https://spacy.io/usage/rule-based-matching) and [Github Spacy issue](https://github.com/explosion/spaCy/issues/2704)

I also want to add other multiple custom entities in the pipeline before our NER along with this regex entity matcher. How could I perform both of these together? Reference examples would really help.

## Your Environment
* Operating System: Windows 10 
* Python Version Used: 3.6.7
* spaCy Version Used: 2.0.12

Thank You in advance. Appreciate the great support!

**Comment by ines** [comment]
Hi! If you want to match regular expressions on the full `doc.text`, it might actually make more sense to use `re.finditer` on the `doc.text` directly, and then call [`doc.char_span`](https://spacy.io/api/doc#char_span) to create a `Span` objects from the start and end character offsets of the matches.

Here's an example (untested, but should work):

```python
import re

def regex_matcher(doc):
    label = "YOUR_LABEL"
    expression = re.compile(r'(\d+\s*mg/m\d* to \d+\smg/m\d)|(\d+ to \d+\s*mg/m\d*)|(\d+\s*mg/m\d*)|(\d+\s*\w+\s*units/m\d*)')
    spans = []
    for match in re.finditer(expression, doc.text):
        start, end = match.span()
        span = doc.char_span(start, end, label=label)
        if span:  # Only add span if it's valid
            spans.append(span)
    # Add spans to the doc.ents
    doc.ents = list(doc.ents) + spans
    return doc
```

Note that `doc.char_span` can only create a `Span` if the start and end incides map to valid tokens. If not, it'll return `None`. If this happens, you might want to print the span and check if it's something you can easily fix, e.g. by customising the tokenization rules.

You might also want to check that there are no duplicates in your matches and `doc.ents`. If you're using spaCy v2.1.4+, you can use the [`util.filter_spans`](https://spacy.io/api/top-level#util.filter_spans) helper for that.

The function above takes a `doc`, matches, adds the entities and returns the `doc` again. This means it's also a valid [custom pipeline component](https://spacy.io/usage/processing-pipelines#custom-components), and you can call `nlp.add_pipe(regex_matcher)` and add it to your pipeline.

**Comment by sohilsshah91** [comment]
Hi @ines  Thanks a lot. I will try this logic now. One more quick question, the reason for getting this as custom entities is I want to create a knowledge graph.  There are 3 of these new custom Entities that I plan to add to current 'en_core_web_sm' library. 1 through the above regex_matcher but other two are list of items eg: list of indications and list of patient types and they need to be added too to as Entities. How could I do multiple NERs with different types of matches utilised in same pipeline?
Thank you again.
Regards,
Sohil Shah

**Comment by ines** [comment]
> How could I do multiple NERs with different types of matches utilised in same pipeline?

You could just rewrite the `regex_matcher` function to match multiple expressions and assign a different label depending on the expression that was matched. For example:

```python
expressions = {
    "LABEL_ONE": re.compile(r'(\d+\s*mg/m\d* to \d+\smg/m\d)|(\d+ to \d+\s*mg/m\d*)|(\d+\s*mg/m\d*)|(\d+\s*\w+\s*units/m\d*)'),
    "LABEL_TWO": re.compile(r'...')  # etc.
}
for label, expression in expressions.items():
    # Match, create spans with label and add to doc.ents here
```

Alternatively, you could also add multiple `regex_matcher` components to the pipeline. The `name` argument on `nlp.add_pipe` lets you assign a custom name.

**Comment by sohilsshah91** [comment]
Hi @ines,
Thanks a lot I will try these. But before that, as soon as I tried to update the library to 2.1.4 (as per suggestion) for the filtering part, I uninstalled and installed again and now I am getting errors like 'No module named spacy.tokens' while doing python -m spacy download en_core_web_sm. I have faced this kind of error before too while installing or updating the library. Even downgrading is not helping. After activating virtual environment is there any way to update or use python commands like python -m .....

Thank You in advance.
